### ISSUE TITLE
Allow for accessing slices of dataset

### ISSUE THREAD

**Issue by VictorSanh** [mosaicml/streaming#214]
## 🚀 Feature Request

Right now, we can use the `StreamingDataset` as a standard iterator or for random access. I want to do sliced random accesses.
Essentially, in the README example, I'd like to do a `sample = dataset[list(range(1337, 1450))]`.

```
from torch.utils.data import DataLoader
from streaming import StreamingDataset

# Remote path where full dataset is persistently stored
remote = 's3://my-bucket/path-to-dataset'

# Local working dir where dataset is cached during operation
local = '/tmp/path-to-dataset'

# Create streaming dataset
dataset = StreamingDataset(local=local, remote=remote, shuffle=True)

# Let's see what is in sample #1337...
sample = dataset[[1337, 1338]] # instead of dataset[1337] and then dataset[1338]
img = sample['image']
cls = sample['class']

# Create PyTorch DataLoader
dataloader = DataLoader(dataset)
```

**Comment by knighton** [comment]
You may like https://github.com/mosaicml/streaming/pull/120, which adds support for passing slices, lists, and ndarrays ;)

Not merged due to higher priorities for now
 
```
sample = dataset[[1337, 1338]] # instead of dataset[1337] and then dataset[1338]
img = sample[0]['image']
cls = sample[0]['class']
```

**Comment by VictorSanh** [comment]
ooooh yes indeed, thanks for the pointer!

do you have an idea of when this is going to be merged?

**Comment by knighton** [comment]
The only blocker: [dataset.py](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py) has been getting too "god object", and one idea is to split it into lower and higher level parts:
* Index (`__getitem__` with sub-dataset concatenation, the shard index, cold shard eviction, etc.)
* StreamingDataset (`__iter__` with shard prefetching, elastic deterministic shuffle, mid-epoch resumption, etc.)

Do we then subclass the inner class to override `__getitem__`, subclass the outer class and pass its `__getitem__` into the inner class internally, or do it outside both of them via `.map()` as an IterDataPipe? All while aiming for simplicity and graduated complexity in the user experience. Presumably `get_item()` would travel with it.

**Comment by karan6181** [comment]
Hi @VictorSanh , PR https://github.com/mosaicml/streaming/pull/120 has been merged. You can now use NumPy style indexing on `StreamingDataset`. Let us know if you face any issues. Thanks!

**Comment by karan6181** [comment]
Closing this issue for now. Please feel free to re-open or create a new one if you face any issues. Thanks!
